source/common/router/router.cc

```diff
@@ -1837,15 +1838,61 @@ bool Filter::convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& do
   const std::string original_host(downstream_headers.getHostValue());
   const std::string original_path(downstream_headers.getPathValue());
   const bool scheme_is_set = (downstream_headers.Scheme() != nullptr);
-  Cleanup restore_original_headers(
-      [&downstream_headers, original_host, original_path, scheme_is_set, scheme_is_http]() {
-        downstream_headers.setHost(original_host);
-        downstream_headers.setPath(original_path);
-        if (scheme_is_set) {
-          downstream_headers.setScheme(scheme_is_http ? Http::Headers::get().SchemeValues.Http
-                                                      : Http::Headers::get().SchemeValues.Https);
+
+  auto saved_headers = Http::RequestHeaderMapImpl::create();
+  std::vector<std::string> headers_to_clear;
+
+  for (const auto& to_preserve :
+       route_entry_->internalRedirectPolicy().responseHeadersToPreserve()) {
+    auto header = Http::LowerCaseString(to_preserve);
+    auto result = upstream_headers.get(header);
+    auto dresult = downstream_headers.get(header);
```

Comment:
I think we can fully qualify downstream_

Reply:
Done.